### ISSUE TITLE
Upgrade to newer Node.js versions, drop Node.js 4 and 7

### ISSUE THREAD

**Issue by generalmimon** [kaitai-io/kaitai_struct#729]
The target [`javascript/nodejs7`](https://github.com/kaitai-io/ci_artifacts/tree/javascript/nodejs7/test_out/javascript) is currently broken in our CI. It is caused by the assertion failure of the [test `ExprOpsParens`](https://github.com/kaitai-io/kaitai_struct_tests/blob/master/formats/expr_ops_parens.ksy) that was recently added. The way that the `./ci-javascript` script breaks is really strange: the only error that gets dumped is that the Ruby XML parser detects [invalid XML](https://travis-ci.org/github/kaitai-io/ci_targets/jobs/670952504#L568-L577). And indeed, looking into the [XML test results](https://github.com/kaitai-io/ci_artifacts/blob/f63aefcb4ce748d19b5e1050e5465ea0b3cb321e/test_out/javascript/test-output-javascript.xml#L78) from Mocha, it just gets terminated after running the ExprOpsParens test.

I can reproduce it on my local environment as well: it seems that doing anything like `assert.strictEqual(2, "2");` (number and string) causes Mocha XUnit builder to break internally, dumping results to the console still works normally.

I asked NodeJS to print the uncaught exception by doing

```javascript
process.on('uncaughtException', function(err) {
  console.log(err);
});
```

and got this:

<details>
  <summary>Mocha XUnit reporter uncaught error</summary>

```
TypeError: value.split is not a function
    at Diff.lineDiff.tokenize (D:\temp\kaitai_struct\tests\node_modules\diff\lib\diff\line.js:19:32)
    at Diff.diff (D:\temp\kaitai_struct\tests\node_modules\diff\lib\diff\base.js:36:39)
    at diffLines (D:\temp\kaitai_struct\tests\node_modules\diff\lib\diff\line.js:44:19)
    at structuredPatch (D:\temp\kaitai_struct\tests\node_modules\diff\lib\patch\create.js:20:83)
    at createTwoFilesPatch (D:\temp\kaitai_struct\tests\node_modules\diff\lib\patch\create.js:126:14)
    at Object.createPatch (D:\temp\kaitai_struct\tests\node_modules\diff\lib\patch\create.js:146:10)
    at unifiedDiff (D:\temp\kaitai_struct\tests\node_modules\mocha\lib\reporters\base.js:418:18)
    at Function.exports.generateDiff (D:\temp\kaitai_struct\tests\node_modules\mocha\lib\reporters\base.js:181:7)
    at XUnit.test (D:\temp\kaitai_struct\tests\node_modules\mocha\lib\reporters\xunit.js:167:23)
    at D:\temp\kaitai_struct\tests\node_modules\mocha\lib\reporters\xunit.js:105:12
    at Array.forEach (<anonymous>)
    at Runner.<anonymous> (D:\temp\kaitai_struct\tests\node_modules\mocha\lib\reporters\xunit.js:104:11)
    at Object.onceWrapper (events.js:286:20)
    at Runner.emit (events.js:203:15)
    at D:\temp\kaitai_struct\tests\node_modules\mocha\lib\runner.js:890:12
    at D:\temp\kaitai_struct\tests\node_modules\mocha\lib\runner.js:762:9
    at next (D:\temp\kaitai_struct\tests\node_modules\mocha\lib\runner.js:362:14)
    at Immediate.<anonymous> (D:\temp\kaitai_struct\tests\node_modules\mocha\lib\runner.js:415:5)
    at runCallback (timers.js:705:18)
    at tryOnImmediate (timers.js:676:5)
    at processImmediate (timers.js:658:5)
```

</details>

It seems like this issue has been already resolved in Mocha (in  https://github.com/mochajs/mocha/commit/1412dc80d87d0479f7f1d60202da2b33c90eb939#diff-d80a115516f0cc2ec0b46a54c60077fa), but it's available only since version 7.0.0 that's no longer compatible with NodeJS 7 we use in CI for testing.

I think we can't really do much to fix this, but I'd suggest another thing. We should probably finally switch to some newer versions of NodeJS. Node.js versions 4 and 7 are pretty much old and deprecated, most projects have already dropped their support for them, and it's probably not worth the trouble to desparately try to save them from extinction. The problem with the deprecated version is that we're pretty much against the rest of the world and when we have some problem with some third party library or even the Node itself, nobody's gonna help us.

@GreyCat, @koczkatamas, @cherue, @jchv  What do you think? I'm just wondering if there's some reason for testing against such old versions, because if isn't, I see no reason for not upgrading.

Looking on the Node.js [release version cycle](https://nodejs.org/en/about/releases/), I'd go for Node.js 8 as "compatibility version" (it has reached End-Of-Life on December 2019 though, but I assume projects potentially using Kaitai Struct do not adapt so quickly) and Node.js 12, being currently the latest LTS version (recommended for production use). I think there's no point supporting any older ones.

**Comment by jchv** [comment]
Node.JS 7 is ancient at this point, last updated in 2017, and I wholly support dropping it. In fact, anything below v10 should be dropped as they are no longer supported upstream.

**Comment by generalmimon** [comment]
@jchv It's incredible how prompt your responses are, as always, tell me how you do it 😃

The nice thing about that change is that it will take just a few seconds to change a few numbers in [travis.yml](https://github.com/kaitai-io/ci_targets/blob/master/.travis.yml#L84-L91), but it will save a lot of trouble in the future.

**Comment by jchv** [comment]
@generalmimon 
> @jchv It's incredible how prompt your responses are, as always, tell me how you do it smiley

Currently, a combination of timing luck, and push notifications :) (GitHub Notifications in Firefox, and the official app on iPhone.)

> The nice thing about that change is that it will take just a few seconds to change a few numbers in [travis.yml](https://github.com/kaitai-io/ci_targets/blob/master/.travis.yml#L84-L91), but it will save a lot of trouble in the future.

Yeah, I think we should go for it. I am hoping for someone with merge access to give the green light, but I suppose either of us can go make a PR right now. (Let me know if you want me to do that.)

**Comment by generalmimon** [comment]
@jchv 
> I am hoping for someone with merge access to give the green light, but I suppose either of us can go make a PR right now. (**Let me know if you want me to do that.**)

Good idea! Could you please do that?

**Comment by jchv** [comment]
Done, https://github.com/kaitai-io/ci_targets/pull/6.

**Comment by cherue** [comment]
I'm in favor of upgrading to the latest version every six months but I'm against dropping Node 4.

Node 4 is the lowest version we can support because that's when Node's `Buffer` was made compatible with `TypedArray`s. If we test against Node 4 and the latest version we can be confident it works in all Node versions in between.

We only have two third party dependencies: iconv-lite and mocha. Iconv-lite is compatible with Node >=0.10.0, it's just mocha that now breaks. But we don't really need mocha, it's basically just used to produce XUnit reports. It could be replaced by a script that runs the tests and puts the result in an XUnit report (this can be achieved by joining some strings, no need for another dependency).

I think the amount of effort of setting up the CI is justified to support practically all versions of Node.

I'll make a PR that replaces mocha with a custom test script.

**Comment by jchv** [comment]
Understood, although I personally think it would be a (minor) maintenance benefit to explicitly drop Node.js v4 support instead (I'm all for keeping dependencies down, but I don't see this one as a huge win personally; what's KaitaiStruct doing implementing something like this? Will it be as robust as Mocha's? etc.)

Node.js v4 was completely EOL'd on April 30th, 2018, and won't be receiving security updates; the main reason I imagine you would come across Node.JS v4.x nowadays is via out-of-date LTS Linux distributions. The situation there is:

- Ubuntu 14.04 shipped v0.10.x (and is out of maintenance anyways.)
- Ubuntu 16.04 shipped v4.x, and is out of maintenance in April, 2021.
- Ubuntu 18.04+ ship v8.x.
- Debian Jessie ships v0.10.x, and is EOL in a couple months (still tentative.)
- Debian Stretch ships v8.x (in backports.)
- Debian Buster (stable) ships v10.
- As far as I can tell, RHEL and CentOS do not ship Node.JS as part of the base repositories at all.
- Versions of Node.JS under v10 are not available on Dockerhub.
- Ubuntu, Debian and CentOS are all covered by NodeSource, which offers supported binaries of stable Node.JS versions for almost everything, even very EOL distributions.

I suggest that in 2021, it will become nearly completely pointless to continue supporting Node.JS v4. If somebody needs support for Node.JS v4 and it breaks, older versions of KaitaiStruct will obviously continue to work, but relieving the necessity to support out of date software versions may allow for better refactoring. As a potential example, the minimum version of JavaScript that an extreme majority of browsers support is already quite far ahead of Node.JS v4 (which, IIRC, doesn't even have full coverage for ES6.)

That having been said, if supporting Node.JS v4 is still desirable, then I understand, feel free to close my pull request. Still, I do think it's healthy to eventually cut the legacy; If an ancient version of Node.JS is OK for someone, I see no reason why an ancient version of KaitaiStruct is more of a problem.

**Comment by GreyCat** [comment]
It's more or less the same discussion as about [Python](https://github.com/kaitai-io/kaitai_struct/issues/695) or [Java](https://github.com/kaitai-io/kaitai_struct/issues/697).

I believe, more or less the consensus on that was that we test LTS versions (normally just 2: the oldest and the newest) which are recognized as still supported by language/platform vendors *plus* rarely we do special exceptions as with Python 2 or Java 7 if there's enough evidence that people still use that.

In this case, we might want to consider that Ubuntu 16.04 LTS (xenial), which is probably one of the most popular systems, [still ships nodejs 4](https://packages.ubuntu.com/xenial/nodejs), and its EOL is April 2021, i.e. at least 1 more year.

That said, obviously, we have no obligations to support these. If everybody thinks that moving forward is a great idea — let's do it.

Talking about JS testing — we still have huge uncovered land of browser compatibility checks...

**Comment by generalmimon** [comment]
So what is the conclusion? As I understood, @cherue and @GreyCat don't want to drop testing against Node.js 4, but nobody objected to dropping Node.js 7. I think this should be the way to go, keep testing against Node.js 4, drop Node.js 7 and start testing against modern versions, i.e. Node.js 12.

I suppose there is a little reason testing against version 10 in addition, 10 and 12 seem to be pretty much compatible (yeah, [there are several changes](https://nodejs.org/tr/blog/uncategorized/10-lts-to-12-lts/), but I think we can assume that if everything works in Node.js 4 and Node.js 12 at the same time, then it will work for all the versions in between - as @cherue suggested). But I don't have much experience with it, if @cherue or @jchv think it's a good idea to test against Node.js 10 too, I'm not against it.

Shall we modify the PR https://github.com/kaitai-io/ci_targets/pull/6 to reach a consensus and then merge it?

**Comment by cherue** [comment]
I think it would be nice to support Node.js 4 forever, just because it's possible. If older versions' Buffer weren't incompatible I'd say the same thing about Node.js 0. But since I'm the only one that feels that way and the general solution is to go by LTS distros for the lowest versions I vote for that instead.

My suggestion is to add a bunch of Node.js versions:
- All active Node.js LTS versions with a comment until when they officially reach EOL, adding new ones as they come out
- All outdated but still relevant Node.js LTS versions with a comment about what distros use them and when those distros become unsupported

That way it's easy to know when to remove old versions (when no active LTS distro has it) and when to add new versions (every April).

The list of versions would currently be:
- Node.js 4 "Argon"
  - Ubuntu 16.04 "Xenial Xerus" (until April 2021)
- Node.js 8 "Carbon"
  - Debian 8 "Jessie" (until June 2022)
  - Ubuntu 18.04 "Bionic Beaver" (until April 2023)
- Node.js 10 "Dubnium"
  - Node.js LTS (until May 2021)
  - Debian 10 "Buster" (until 2024)
- Node.js 12 "Erbium"
  - Node.js LTS (until May 2022)

**Comment by generalmimon** [comment]
Based on the above consensus, I removed Node.js 4 and Node.js 8 from our test CI matrix. Both have been end-of-life for a long time, Node.js 4 since April 2018 and Node.js 8 since December 2019 (see https://endoflife.date/nodejs). And they are not part of any supported Ubuntu or Debian distribution.

At the moment, the oldest supported version of Ubuntu is 20.04 "Focal Fossa" (see https://endoflife.date/ubuntu), which is supported until April 2025 and provides Node.js 10 (see https://packages.ubuntu.com/search?keywords=nodejs&exact=1 or https://launchpad.net/ubuntu/+source/nodejs).

The oldest supported Debian is 11 "Bullseye" (see https://endoflife.date/debian), which is supported under [Debian LTS](https://wiki.debian.org/LTS) until August 2026 and provides Node.js 12 (see https://packages.debian.org/search?keywords=nodejs&exact=1). Debian LTS for version 10 "Buster" ended in June 2024, it provided Node.js 10.

So for now I would test on the following versions:

* Node.js 10 (because Ubuntu 20.04 "Focal Fossa" is supported until April 2025, then we can drop it)
* Node.js 12 (because it's included in Ubuntu 22.04 "Jammy Jellyfish", which is supported until April 2027; as mentioned above, it is also in Debian 11 "Bullseye" supported under Debian LTS until August 2026)
* Node.js 18 (oldest supported LTS version)
* Node.js 22 (latest LTS version)
